### ISSUE TITLE
Should we recommend CURIEs for use as identifiers?

### ISSUE THREAD

**Issue by reece** [ga4gh/vr-spec#32]
## Background 

VMC currently uses [Identifier](https://docs.google.com/document/d/12E8WbQlvfZWk5NrxwLytmympPby6vsv60RxCeD5wc1E/edit#heading=h.kj21ornkb4u7) objects that consist of a `namespace` and `accession`.  These are be (nearly) equivalently represented as JSON (e.g., `{"accession": "GA_0123abcd", "namespace": "VMC"}`) and as a string (e.g., `"VMC:GA_0123abcd"`). A VMC `accession` is constructed as `prefix + "_" + digest`, where prefix is obtained from a mapping from types to prefixes, such as `"VMC:GA_0123abcd"` for Alleles, `"VMC:GL_0123abcd"` for Locations, etc.

The VMC string format is inspired by long-standing conventions of colon-separated `namespace`:`key` pairs, which is also consistent with the [W3C CURIE](https://www.w3.org/TR/curie/) syntax and semantics. It is anticipated that future VMC specs will adopt CURIEs as the Identifier format rather than defining its own syntax.

Briefly, a CURIE adds additional meaning to the current `namespace:accession` scheme (`prefix:reference` in CURIE parlance): `prefix` is an alias for a base IRI/URI, and `reference` is a (possibly) relative IRI/URI. A fully-qualified IRI/URI is constructed by concatenating the `prefix` and `reference`, separated by a `/`. In other words, CURIEs serve to provide a convenient shorthand for an object.

## Question

What format should the VMC Identifier have?

**Comment by reece** [comment]
There's nothing wrong with the current Identifier AFAICT.

However, as we move to CURIEs, I think there are some advantages (and some disadvantages) to treating the `accession` (the CURIE `reference`) as a relative IRI (path). In particular, I propose that we replace the accession prefix (GA, etc) with a relative path prefix instead. For example, `"VMC:GA_0123abcd"` would become `"VMC:allele/0123abcd"`.

## Discussion

**The primary advantage is simplified client and server implementation.** Under the current scheme, translating an identifier like `"VMC:GA_0123abcd"` into a URL like`https://api.somewhere.org/allele/GA_0123abcd` requires that either the client or the server peek into the id syntax in order to assess what type of object is requested, which violates the intended opaqueness of the id. This practice entrenches an unfortunate coupling between the `allele` path component and the `GA` prefix. In contrast, using a relative path (e.g., `allele/0123abcd`) as a `reference` provides a natural translation from CURIE to URI forms (by simple concatenation).

**A secondary advantage is extensibility.** Current two-character prefixes (GS, GL, GA, GH, GG, etc) have very limited range. In contrast, permitting relative paths enables much more flexibility.

**Relative paths are less familiar and more verbose than prefixed identifiers.** It is anticipated that more users are familiar with ids like `GA_0123abcd` than `allele/0123abcd`. Users might be confused by `references` that are relative paths rather than prefixes., or might prefer the prefix syntax for aesthetic, stylistic, or familiarity reasons.

**Comment by reece** [comment]
@jmcmurry -- I would appreciate your opinion on this question. Thanks!

**Comment by jmcmurry** [comment]
All good points but there are always tradeoffs. There's no technical prohibition, but I have a strong suspicion that a slash mid ID will result in parsing errors by others when doing curie expansion/contraction.

Your best bet is probably either VMC:GA0123abcd or better yet just VMC:123abcd if the performance is not compromised. You can always redirect to a nested URL if you want to whether 

https://api.somewhere.org/allele/123abcd or 
https://api.somewhere.org/allele/GA0123abcd 

or no nesting:
https://api.somewhere.org/123abcd (requires guarding against collisions across types, and requires type lookups)
https://api.somewhere.org/GA0123abcd

As long as the biological types are durable, coarse, useful and undisputed, baking type into local ID is ok.

**Comment by larrybabb** [comment]
@Reece Do we have to have a single VMC namespace? I understand the desire to do this, but maybe it would be okay to consider the idea of something like
VMCGA (or VRGA) - allele
VMCGL (or VRGL) - location
VMCGS (or VRGS) - sequence
etc..
We shouldn't be too married to the VMC acronym. The GA4GH refers to us as the VR group and the VMC (variant modeling collaboration/consortium) is a bit outdated and confusing - possibly. 
I'm thinking that it would be reasonable to have several namespaces which target the various classes of things that are digested in the namespace, essentially using the namespace to prevent computed identifier collision. I've never been a big fan of manipulating (prefixing) identifiers in order to jam them into a single namespace, but I get that we also don't want an explosion of namespaces to manage too.

ClinVar and others do the prefixing approach (VCV000x, RCV000x, SCV000x, etc..) which keep all like identifiers/accessions in their one authorizing identifier system. So, I think we should stay with that approach as well.  We could expand to a 3 latter prefix if you are concerned about running out of options.
Not sure if any of that is helpful, but I thought I'd share some thoughts for what its worth.

**Comment by larrybabb** [comment]
@reece btw why do we prefix our prefixes with "G"? (is it for "generated"?)

**Comment by ahwagner** [comment]
I completely agree with @jmcmurry here; my immediate concerns about switching are precisely captured by her comments above.

I would like to add that a third alternative–dropping the underscore–is preferable in my view, and is in line with @larrybabb's suggestion re: ClinVar practices, as well as those that we see from Ensembl (genes/transcripts/proteins/etc), the ClinGen Allele Registry, dbSNP, and others. Of course, there's nothing wrong with keeping the underscore either (RefSeq style), with the minor exception of the additional bulk from the extra character in ID strings.

**Comment by reece** [comment]
@jmcmurry: Thanks for responding so quickly. You're the second person to suspect that CURIE handling in the wild might not support relative paths. That's a shame because rel refs would be useful and because spec clearly says they're permitted. :-( Any idea how we could get a handle on whether this is a widespread problem?

Re: Dropping the prefix (@jmcmurry): Dropping the prefix would require backends to search for objects in each type (or keep a map of these types). Digest collisions between types is not a concern.

Re: Underscore (@jmcmurry, @ahwagner). This is tied to the id prefix discussion. Aesthetically, I also prefer without. I added the underscore to prevent parsing ambiguity between the id prefix and digest. For example, is GAA012abc an allele with id A012abc, or a new type with prefix GAA and id 012abc? Using a rel path in the CURIE reference would be a standards-based way to enable us to distinguish id type prefix and digest.

Re: Prefix: The G prefix was originally for "global" or "GA4GH", depending on when you asked me.

Re: New prefixes (@larrybabb): The conventional interpretation of the namespace/prefix is as the data source. (See https://www.ebi.ac.uk/miriam/main/collections for examples.) Sometimes namespace includes sub-schemas. I don't know of any examples where namespace incorporates a single data type. There are two problems I see with including type in the prefix. ① It creates more burden for developers because they should provide CURIE prefix-to-uri maps, and doing so for every prefix will be tedious. ② Expanding types would require making code aware of associated new prefixes.

Re: Namespace: I'm not wedded to "VMC". But, let's stick with "VMC" as a placeholder for now, then then propose a switch to "GA4GH" closer to ratification. (IMO, "VR" offers no advantage over "VMC".)

**Comment by reece** [comment]
So, here are options so far:

* `namespace:prefix_digest` (e.g., `GA4GH:GA_0123abcd`). Current. No problems.
* `namespace:prefixdigest` (e.g., `GA4GH:GA0123abcd`). Same as above, but drops the underscore. Since we know that implementations will parse the id, this really boils down to aesthetics and preference for telling people to "take the first two chars" or "take chars up to underscore".
* `namespace:type/digest` (e.g., `GA4GH:allele/0123abcd`), à la CURIE. Pros and cons in first comment in thread. Basically, a standards-based equivalent to `GA4GH:GA_0123abcd` with easier implementation. 
* `namespaceprefix:digest` (e.g., `GA4GHGA:0123abcd`). Theoretically workable, but likely to not scale well as new types are created.

**Comment by jmcmurry** [comment]
Option | My take
---|--
namespace:prefixdigest (e.g., GA4GH:GA0123abcd). | This is best; it is also fine to just say take all of the alpha chars up to the first number digit.
namespace:prefix_digest (e.g., GA4GH:GA_0123abcd).|Same as above, but has the underscore. In Monarch, we have observed edge cases in the wild in which local ID underscores cause trouble due to lossiness about the understanding of what the prefix is*
namespace:type/digest (e.g., GA4GH:allele/0123abcd), à la CURIE. | I don't think this will be as robust as you'd like in others hands. In the wild, there's already a lot of variation in the way people represent prefixes, if the local ID has room for differing interpretations, it will likely complicate integration even more. Specifically, you'll end up with equivalents you can't easily collapse because they require more ID munging to discover**
namespaceprefix:digest (e.g., GA4GHGA:0123abcd).  | If you have (or will have) a type-agnostic resolving path, embedding your ``type flag`` in the prefix can break the 1:1 mapping between resolving path and prefix which is one of the strongest advantages of the CURIE.

If the option exists to have the local IDs be unique even in the absence of a ``type flag``, that is ideal. It is definitely harder to implement as there may be various bodies that need to mint IDs in collaboration. It is also harder for systems that have to work with billions of IDs, so it is not for everyone. If the number of IDs you need is small (millions not billions) you can achieve uniqueness by reserving blocks of IDs for each type. However, obviously I would caution against attempting this unless you are 100% sure you can guarantee uniqueness in the long run (eg. you have control over the full lifecycle of IDs).

Trivia:
'* Underscores are a pervasive stand-in for colons; the underscore to colon works fine, but then you run into trouble on the round trip; which of the two underscores do you re-convert to a colon?
'** If you imagine you have the full URL, it is pretty likely that someone will assume that the fragment after the final slash is the thing that needs to be contracted to a prefix.

**Comment by reece** [comment]
Based on advice from @jmcmurry and discussions with @ahwagner , we're going with option 1 in Julie's table (2-char prefix, no underscore).

**Comment by reece** [comment]
There's been some post-close chatter on this topic. I'm reopening to solicit 11th-hour feedback.

Here are real examples to try on for size:
`ga4gh:VA/_7JERW4UiCUgH8oaro2QmcpTwRBUFnjU`
`ga4gh:VA__7JERW4UiCUgH8oaro2QmcpTwRBUFnjU`
`ga4gh:VA_7JERW4UiCUgH8oaro2QmcpTwRBUFnjU`
(This digest starts with an underscore.)

Part of this decision is whether the identifier is supposed to be totally opaque, in which case we might as well drop the type prefix altogether, or, instead, it's translucent (encodes type), in which case we want to make the type useful and extensible. / does that best since it's guaranteed to not be part of the digest. (We chose base64url precisely so that the digest could be used in a path.)

My vote is to switch to /-delimited. The spec already says that we use CURIEs, and / is definitely legal in a CURIE. The hypothesized failures of some tools to handle CURIEs correctly should not keep us from doing so.

**Comment by ahwagner** [comment]
What about a GA4GH:VA:<digest> structure? Solves the underscore problem and provides us extensibility, without introducing potential human errors that @jmcmurry described for /-delimited. While I appreciate the technical elegance of the / delimiter, I agree with her assessment that it won't be robust in the field.

**Comment by reece** [comment]
What's an example use case or tool that would be adversely impacted by permitting the use of CURIEs (with slashes)?

This decision is asking us to consider NOT adopting a standard because we fear that t won't work with some tools.

I'm okay with that unfortunate consequence in the name of pragmatism in principle, but let's base it on real data regarding problems and weigh those problems against the lost value of using the CURIE standard.

**Comment by ahwagner** [comment]
Truthfully, the use of colons OR slashes are problematic, though I think slashes are more so. Use of these will cause conflicts (or require special handling) when used as query strings in systems that harbor variants under multiple classification schemes and no underlying logic to handle the proposed VR identifiers specifically. Slashes have the additional complexity of being used as path delimiters, both in URLs and local filesystems, requiring 

For example, in VICC our webtool uses Kibana on top of ElasticSearch, a common prototyping framework for document searching. ':' and '/' are reserved characters for lucene search, and so these queries cause problems:

https://search.cancervariants.org/#GA4GH:allele/0123abcd
https://search.cancervariants.org/#GA4GH:allele:0123abcd

Notably, an identifier of the type VA_\<digest> or GAVA_\<digest> doesn't have these issues. The potential conflicts of multiple '_' appearances in such a scheme is an argument for adopting the TRUNC512 hex encoding as well (for reference, this discussion is in #58).

In summary, I propose we adopt the following scheme: 

\<org-prefix>:\<type>_\<digest>
e.g. GA4GH:VA_0123abcd

**and** we follow refget and use hex encodings to avoid the potential parsing conflicts @reece mentioned [above](https://github.com/ga4gh/vr-schema/issues/32#issuecomment-497500615).

This has the advantages of:
* closer alignment to refget
* same structure used by Entrez
* minimal lucene search conflicts (same conflicts need addressing as with HGVS queries)
* CURIE syntax
* in line with best practices as described by @jmcmurry 
* minimal difference from our current implementation

**Comment by reece** [comment]
Based on #56, and regardless of terminology names there, I think we're leaning toward identifying objects with unconstrained strings and RECOMMENDING a structured string to provide external references.

## What's the question in this issue? 

The question is whether we adopt CURIEs as the id/identifier format. And, if so, provide substructure within the CURIE for GA4GH-specific meaning.

If we're not willing to adopt the full meaning of CURIEs, then we shouldn't use that name. Instead, we need to define a home-grown structure. (And it will likely look like a CURIE but not actually be one.)


## How we make this decision

The goal for this decision is an expression that is parseable reliably by software, and preferably in diverse systems. These expressions are not for humans therefore the decision does NOT involve aesthetics to humans.

I will presume that we agree that the decision must be based on technical correctness and non-technical concerns that impact multiple systems, not specific components. I'll also presume that when implementations fail because they or their components don't use (or misuse) standards, the default is that they should adapt. 


## Decisions to make

### Decision #1: CURIEs or home-grown colon-delimited syntax for id structure?
As authors of a standard, we should use existing standards wherever possible. And, we should expect that others will use standards correctly. 

CURIEs are designed for our use case and we should use them. That means *permitting* slashes, and therefore that consumers must tolerate slashes (and colons, and anything else that's in the CURIE spec).

CURIEs consist of a `prefix` and `reference` (W3C terminology), which we colloquially call "namespace" and "accession", and [McMurry 2017] calls "namespace" and "local id". I'll use W3C here because I want to be clear that I'm using the terms as they define them.

### Decision #2: What CURIE `prefix`?
The CURIE spec places no requirements on the meaning of `prefix`. I would like VR to RECOMMEND using identifiers.org as the source of acceptable prefixes. identifiers.org uses lowercase. "ga4gh" is the answer here.

### Decision #3: What structure for `reference`?
For `reference`, I think we all agree that we want a translucent string, i.e., with an embedded type identifier of some sort. (A transparent string, like an HGVS expression, is not an option. And I think we all agree that a truly opaque string such as a raw digest is also not an option.)

If a `reference` string has type, we're putting it there to be parsed in some contexts. Our question now is what mechanism is most amenable to parsing?

The options I see:
* fixed-width, two characters: Pros: Easy to explain; Cons: non-standard and not extensible.
* underscore- and colon- delimited: Pros: familiar; Cons: underscore is a valid member of the `pchar` characters for URI paths (RFC3986), so simultaneously conflicts with path content and precludes the use of URI libraries to parse that concept from the expression. (I'm not saying that this scheme doesn't work or that it's even wrong in any absolute sense.)
* slash-delimited: Pros: Complies with CURIE spec and therefore enables the use of tools; easily extends to subtypes; Cons: Code that doesn't expect CURIEs might be confused.


## Response to @ahwagner's implementation issue
RFC3986 permits colons and slashes in fragment identifiers. The example URLs are completely legal. I don't see these URL examples themselves as causing a technical problem by software that obeys standards.

The Lucene issue is solved by correctly escaping reserved characters during querying. This is a common question and many Google responses cover it. Moreover, *all* software that passes messages must deal with escaping and quoting to prevent message content from conflicting with message structure. So, unless I'm missing something, I would expect a variant of this issue with all software dependencies.


If anything, these examples support the notion that standards prevent confusion and that we shouldn't make up our own scheme.

**Comment by jmcmurry** [comment]
I advise omitting the extra delimiters unless there is a very strong reason not to. While the spec will technically support all kinds of permutations, Murphy's law will prevail in the wild: someone out there will botch what can be botched. Lowercase prefix is fine.

Eg. ga4gh:**VA**7JERW4UiCUgH8oaro2QmcpTwRBUFnjU

If you define the ``type flag`` as the alpha characters up to the first numeric digit, parsers should not have any trouble. This has the added benefit of being shorter and robust to ``type flags`` of varying lengths.

**Comment by reece** [comment]
Thanks @jmcmurry. Your ears must have been tingling: we were just talking about this topic and asking you for clarification. In https://github.com/ga4gh/vr-spec/issues/32#issuecomment-479179902, you wrote "... you'll end up with equivalents you can't easily collapse because they require more ID munging to discover**" and "** If you imagine you have the full URL, it is pretty likely that someone will assume that the fragment after the final slash is the thing that needs to be contracted to a prefix."

I think what you mean is that if someone sees `ga4gh:VA/blah`, they're going to parse out `blah`, then want prepend `VA` to it, which just creates confusion. Is that right?

**Comment by jmcmurry** [comment]
The problem is the the two-delim versions of any compact ID -- when botched in the wild -- are lossy. 
GA4GH:VA_0123abcd can get (improperly) converted to GA4GH:VA:0123abcd. Then what? How do you convert it back? Is the local part VA:0123abcd? or 0123abcd? Similarly, what is the prefix and what does it expand to? 

Sure, the underscore is not an insurmountable. If people read the fine print in documentation, all works; there are plenty of shops where this is done. But the edge cases are so easily avoided, hardly seems worth the trouble (unless I'm missing something fundamental about the use case).

**Comment by reece** [comment]
I'm not inclined to base technical decisions on the myriad ways that people might botch up identifiers. 

Honestly, cases like these make me even more convinced of the benefit of using a standard syntax for identifiers so that people eventually stop trying to rewrite, "clean up", or otherwise beautify ids.

We're just not going to be able to prevent people from shooting themselves in the foot. If easily avoided, sure, but I don't see it as our problem to design a spec to protect people from themselves, and especially not if we have to forego other benefits to do so.

Some of those benefits:
* Use the `prefix` (namespace) as a full base URL to an object *without* server side munging. That is, `ga4gh:VA/id` becomes `https://somewhere/VA/id` on the client, rather than requiring client or server to rewrite VA_id as the relative path. That makes life a lot easier for clients and servers.
* Clearer semantics. IMO, it's much cleaner to tell people who want the type to "take the first path element" (for which they could use a url parsing library) rather than "take the first 2 chars" or "take the chars up a delimiter".
* More extensible. By using /, we can construct paths any way we want. For example, a subtype could be `allele/snv/id`. 

For me, these are all *known* benefits of using the full power of CURIEs, and, conversely, losses if we don't do so. None of them make-or-break the spec, but I do have a hard time justifying the loss of these benefits against the possibility that people might shoot their own feet.

**Comment by reece** [comment]
Summary of today's call re: identifiers:

Larry proposed that we keep the currently implemented /-delimited paths and revisit in two weeks. In that time, we might have real data about the harms of using paths and can better decide between the costs and benefits.

@jmcmurry @larrybabb @ahwagner @rrfreimuth @mbrush @javild

**Comment by ahwagner** [comment]
I think we need to get the input from other Driver Projects here–I can tell you that for the VICC this is non-trivial for our current implementation (as mentioned [above](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-497729367)), and may affect our ability to spin something up in our proposed timeframe. @melissacline, @ronakypatel could we get your input on this?

I can say that this topic has also come up in Cloud, quite recently: https://github.com/ga4gh/data-repository-service-schemas/issues/263 and https://github.com/ga4gh/data-repository-service-schemas/pull/271.

Ultimately, they agreed to keep `/` in IDs only if it was URL-encoded in the ID, which is obviously not ideal.

**Comment by ronakypatel** [comment]
We might not face this issue as we are not trying to dereference identifier for sequence. Although we may face an issue with dereferencing the sequence identifier (refget). I am not sure why you want to use slash in the identifier (value), let it be part of the prefix.  

namespace: ga4gh.endpoint.org/allele
id: 0123abcd

After thinking I think putting a slash in the id does not feel like an appropriate thing to do. Probable reason why base64URL uses "-" and "_" instead of **"+"** and **"/"**, making it URL safe.

**Comment by ahwagner** [comment]
I have been considering the potential gains of the current `/`-delimited embedded type ids [described above](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-498393382):

> * Use the `prefix` (namespace) as a full base URL to an object _without_ server side munging. That is, `ga4gh:VA/id` becomes `https://somewhere/VA/id` on the client, rather than requiring client or server to rewrite VA_id as the relative path. That makes life a lot easier for clients and servers.
> * Clearer semantics. IMO, it's much cleaner to tell people who want the type to "take the first path element" (for which they could use a url parsing library) rather than "take the first 2 chars" or "take the chars up a delimiter".
> * More extensible. By using /, we can construct paths any way we want. For example, a subtype could be `allele/snv/id`.

I am concerned that in practice these advantages will rarely be realized, while potentially confounding adoption. I think that this strategy will unnecessarily slow or prohibit adoption of the specification for 3 reasons:

# Reason 1: Lack of Motivating Examples

AFAIK there are not other variant representation specifications that have a `/`-delimited id structure. Are there other biomedical data specifications that do? If we’re going to promote this as a _standard_ (versus an academic exercise) I think it will be necessary to point to example implementations that have used **and benefited from** ids with prescribed pathing structures built in via `/`-delimiters. In an ideal GA4GH scenario, these implementations would be (at a minimum) by Driver Projects.

I provided the [above exemplar](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-497729367) from the VICC Driver Project to demonstrate how this structure (`/`-delimiters in an id) may cause problems. A reasonably simple strategy for handling these has been identified in this particular case, and will be documented in the VICC implementation in the event this issue resolves with a decision to promote the `/`-delimited identifiers. However, it is notable that the use of the `/`-delimited typed identifier **did not realize the above cited benefits**, and is more challenging to handle than a similar CURIE without embedded `/` characters. This is because the VICC system treats ids as opaque string objects for searching (alongside other variant representations), and lacking an endpoint to specifically handle the VR object structure does not realize any of these theoretical gains.

Similarly, the ClinGen Allele Registry will not be storing the computed ids, only returning them–thus also not realizing any of the above benefits in the near-term (i.e. before PRC submission), if at all.

Ultimately, I think that this process illustrates a fundamental problem: `/`-delimiters create a need for implementations to modify or create strategies to handle a rare id structure, and we lack example implementations that have asked for or intend to use this structure for the theoretical benefit. As this issue was also recently raised in the [Cloud workstream](https://github.com/ga4gh/data-repository-service-schemas/issues/263), it stands to reason that it will catch the attention of the PRC. If we push forward anyway, I think we’ll need concrete evidence that the cited gains are substantial and likely to be realized.

# Reason 2: Intrinsic Pathing Advantages Tied to REST APIs

The benefits of identifiers with built-in paths are inherently targeted at REST APIs, but our specification otherwise does not make requirements for API behaviors or endpoints. In the VR 1.0 specification, we do make some requirements about the behavior of implementations, but it is limited to the structure of the VR objects themselves and the methods for computing identifiers on them. By extending our data structures *within* ids (e.g. the above `allele/snv/id`) for the purpose of opaque/translucent identifier evaluation, we are essentially defining endpoint structure and API behavior if implementations wish to realize the primary benefit of the `/`-delimiter design choice.

# Reason 3: Downstream munging difficulties

As a variant representation specification, these ids will be used in applications beyond inter-system communications, such as in variation dataset analysis. Consequently, it is certain that data munging efforts will take place for analyzing variation data from one or more datastores. @jmcmurry [points out](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-498355432) that while `/`-delimited ids are problematic, so are `_`-delimited ids, because they (and other identifiers with multiple delimiters) require special care to reconstruct. 

# Are Embedded Types Needed?

@ronakypatel has [also pointed out](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-498430182) that there is no apparent reason that the corresponding object type need be embedded in the id itself. In fact, the [ideal scenario](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-479179902) would be to forgo having the type embedded in the CURIE at all.

As the computed identifier is a value hash for shortcutting queries for VR objects, it stands to reason that:
1) if a queried typed id does not exist in a datastore, the id-embedded object type is of no value
2) if a queried typed id does exist in a datastore, the linked object contains the type information
3) if all ids share the same structure (as untyped ids), they could all be simultaneously queried 

So, as far as I can tell, there is no consequence to dropping the embedded type entirely. 

# Proposal
I suggest we adopt @jmcmurry’s suggestion to do away with `type` flags altogether as the id digests are [unique enough without them](https://docs.google.com/document/d/12E8WbQlvfZWk5NrxwLytmympPby6vsv60RxCeD5wc1E/edit#heading=h.2nwv6g36f686). As a side-note, I would also like to add `.vr` to the namespace for this specification, as it is probable that the GA4GH will define other CURIE ids that are outside of the vr scope. As a result of this, the computed ID for an allele would be:

**Namespace**: ga4gh.vr
**CURIE**: ga4gh.vr:\<digest>

In summary, by dropping type from the CURIE entirely the proposed opaque representation would:
* remain a CURIE, and thus in-line with identifier best practices
* maintain or improve upon all of the cited advantages of a `/`-delimited embedded-type id
  * Easy adoption by servers/clients: ga4gh.vr:id becomes https://somewhere.org/ga4gh-vr/id
  * Easiest semantics. Ids are opaque, no parsing/munging needed.
  * Fully extensible. Type isn’t embedded, so an id can be of any current or future type!
* reduce potential data munging errors in downstream efforts
* use less space (or equivalent space with less implementation effort) with minimally sized strings
* not implicitly define REST API
* be identical to [refget Checksum Algorithm](https://samtools.github.io/hts-specs/refget.html#refget-checksum-algorithm) after re-encoding to base64url, simplifying interchange between formats
  * future version of refget [could adopt this](https://github.com/ga4gh/vr-spec/issues/58#issuecomment-499241127)

Thoughts on this from the collective?

**Comment by cyenyxe** [comment]
I like the style of the last proposal `ga4gh.vr:<digest>`. It gets rid of most issues mentioned in the thread, it is still easy to convert to a URL, and makes clear the sub-namespace within GA4GH.

My only question is if CURIEs, as proposed by the W3C, are supposed to represent a particular resource type or are completely arbitrary. My impression from reading the recommendation is the latter, in which case I would be okay with dropping the type; otherwise I would also be okay with something like `ga4gh.vr.allele:<digest>`

**Comment by ronakypatel** [comment]
I think I broadly agree with @ahwagner @cyenyxe. Just to augment what they said: storing the id with namespace takes extra bytes, without much value. It might be trivial for a few million variants but it is significant for a few billion variants.

**Comment by larrybabb** [comment]
During the VR call today the discussion about the use of prefixes (i.e. SQ, VR, VT, etc...) in the computed identifiers came up again. As well as, the contention over the use of the slash to delimit it with respect to the computed digest value itself.  Here some of what was presented by me.

### Do we need the prefixes in the final computed identifier at all?
Is there a strong argument to why this is needed. Is there a stronger argument for dropping the prefixes all together and simply having the computed digest be the sole id portion of the identifier and the namespace prefix of the CURIE be set to ga4gh-vr. 
Example: ga4gh-vr:Zp8e9tUGfwiTRfIIij7INFQKa_Hc4ep9

**Comment by jmcmurry** [comment]
If you take no other advice from me, that is OK but please please do not embed the type flag in the prefix. My comment[ from april above](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-479179902) still stands.

**Comment by susheel** [comment]
The consensus from yesterday's [GA4GH Cloud call](https://docs.google.com/document/d/1MosGXod2clTB0-JQcD9WG7jYm6BvM5aV41TSF20UqPo/edit#heading=h.vx4stb544m2f) was to disallow slashes and only allow [IETF RFC 3986 § 2.3](https://tools.ietf.org/html/rfc3986#section-2.3) unreserved chars [A-Za-z0-9.-_~]

Hope that helps.

**Comment by ahwagner** [comment]
@jmcmurry your point in https://github.com/ga4gh/vr-spec/issues/32#issuecomment-500623542 is well taken:

> If you take no other advice from me, that is OK but please please do not embed the type flag in the prefix if you include it at all. My comment from april above still stands.

And yes, I agree that the type (if used at all) does not belong in the namespace. I'd like to clarify that the [above proposed](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-499330677) `ga4gh.vr` namespace is for the VR specification, not any specific data type (I believe this is what @larrybabb also intended in the following `ga4gh-vr` proposal from https://github.com/ga4gh/vr-spec/issues/32#issuecomment-500564929). This is to avoid conflict with other standards that ga4gh may build and represent as CURIEs.

**Comment by larrybabb** [comment]
Regarding @susheel's comment above. Could you clarify that the consensus decision you made regarding slashes is related to naming concepts related to `files` and not necessarily about `Ids`? I read through most of the issues you referenced and it seemed to be mostly about dealing with the representation of hierarchical file system structures and how to flatten them without incurring the cost of including the slash in the names.

If this consensus has a direct impact on constraints related to `ids` or `identifiers` (at least the value that is after the namespace) then it would be helpful to know.

I do think the general concern about slashes may apply generally, but I think I lost the direct line of reasoning related to our context.

**Comment by ahwagner** [comment]
Today we finished discussion on this issue, and have come to the consensus to adopt @reece's proposal [above](https://github.com/ga4gh/vr-spec/issues/32#issuecomment-497500615):

`ga4gh:<type>/<digest>`

However, we REQUIRE that only the `<digest>` be used to generate a computed `Id`. This change is being introduced to allow for future backwards-compatible changes to the id structure.

The use of `ga4gh` as the namespace is being discussed with GA4GH Work Stream leadership.

As the strongest dissenting vote to adopting this proposal, I want to be clear that there are theoretical advantages to it. My primary objection is that the observed drawbacks of implementing it will outweigh these potential advantages, which I am concerned will not be realized.

In recognition of approaching deadlines for PRC submission, further discussion on this topic will be considered only after submission on June 24th, 2019.
